[terra-search-field] adding onInput prop

BenBoersma opened the pull request:
### Summary
Fixes https://github.com/cerner/terra-core/issues/2727
Adding an onInput prop so that users can programmatically set the value of search field, then trigger an onInput event to have the component update the internal variable to the new value.

Please add your name to the [CONTRIBUTORS.md] file. Adding your name to the [CONTRIBUTORS.md] file signifies agreement to all rights and reservations provided by the [License].

Thanks for contributing to Terra.
@cerner/terra

[CONTRIBUTORS.md]: ../CONTRIBUTORS.md
[License]: ../LICENSE

emilyrohrbough commented on packages/terra-search-field/src/SearchField.jsx:
```diff
@@ -107,15 +111,15 @@ class SearchField extends React.Component {
     this.handleKeyDown = this.handleKeyDown.bind(this);
+    this.handleInput = this.handleInput.bind(this);
     this.setInputRef = this.setInputRef.bind(this);
-    this.updateSearchText = this.updateSearchText.bind(this);
```
I think we still want this to be bound

BenBoersma replied:
Fixed https://github.com/cerner/terra-core/pull/2800/commits/7f1b23a3ff00608c4dcf47ceae24a47d2b94db4d#diff-105fddc0e7a42eb008cb5c09faad5222R116

emilyrohrbough commented on packages/terra-search-field/src/SearchField.jsx:
```diff
@@ -60,6 +60,11 @@ const propTypes = {
   onChange: PropTypes.func,
 
+  /**
+   * Function to trigger when user inputs an value.
```
```suggestion
   * Function to trigger when user inputs a value.
```

emilyrohrbough replied:
Can we add info on the callback params consumers should expect?

BenBoersma replied:
Fixed https://github.com/cerner/terra-core/pull/2800/commits/7f1b23a3ff00608c4dcf47ceae24a47d2b94db4d#diff-105fddc0e7a42eb008cb5c09faad5222R64

emilyrohrbough commented on packages/terra-search-field/src/SearchField.jsx:
```diff
@@ -107,15 +111,15 @@ class SearchField extends React.Component {
 
     this.searchTimeout = null;
-    this.state = { searchText: this.props.defaultValue || this.props.value };
+    this.searchText = this.props.defaultValue || this.props.value;
```
Why did we remove this from state?

BenBoersma replied:
Since we now have both a handleInput and handleTextChange, when you type a value into the input they both fire. Since setState is asynchronous this meant most of the time both would run the updateSearchText forceUpdate calls.

emilyrohrbough replied:
> Since setState is asynchronous this meant most of the time both would run the updateSearchText forceUpdate calls.

what do you mean?

BenBoersma replied:
Typing the value into the components results in onInput and onChange being called in quick succession of each other. Fast enough that when I left it as a state and put in console messages to follow logic flow, I kept getting multiple setState calls on each keystroke.

emilyrohrbough commented on packages/terra-search-field/src/terra-dev-site/doc/search-field/SearchField.1.doc.jsx:
```diff
@@ -92,6 +94,11 @@ const DocPage = () => (
+      {
+        title: 'Search Field with setting a value programmatically',
+        example: <SearchFieldProgrammaticSet />,
+      source: SearchFieldProgrammaticSetSrc,
```
do we need a first class example for consumers?

BenBoersma replied:
Since there is a bit of complexity getting the expected functionality out of programmatically setting a value because of the need to create a custom 'input' event dispatch, @dkasper-was-taken and I wanted an example.

StephenEsser commented on packages/terra-search-field/src/SearchField.jsx:
```diff
@@ -258,6 +275,7 @@ class SearchField extends React.Component {
             ariaLabel={inputText}
             aria-disabled={isDisabled}
             onKeyDown={this.handleKeyDown}
+            onInput={this.handleInput}
```
We may want to consider adding onInput as first class prop to the terra-form-input/Field so we don't ever lose track of this. Some teams may use dictation for regular inputs.

BenBoersma replied:
I agree with this, but don't think it should be part of this PR. Will log an issue.

StephenEsser commented on packages/terra-search-field/src/SearchField.jsx:
```diff
@@ -162,8 +167,19 @@ class SearchField extends React.Component {
+    this.updateSearchText(textValue);
+
+    if (this.props.onInput) {
+      this.props.onInput(event, textValue);
```
Sending back two parameters seems a little strange. We'll want to be as consistent as we can with these callback parameters. If we do decide to add this to the input field we'll have to add additional function logic to return the two parameters as opposed to just passing the function straight through to the input like we do the onChange.

https://github.com/cerner/terra-core/blob/7f1b23a3ff00608c4dcf47ceae24a47d2b94db4d/packages/terra-form-input/src/Input.jsx#L161-L163

Personally, I would keep this native. It's passive to add it in the future, but non-passive to take it away.

```jsx
    if (this.props.onInput) {
      this.props.onInput(event); // Only returning the event.
    }
```

BenBoersma replied:
Should I do the same thing to our onChange handling in search-field as well? Because I was keeping it consistent with how the 'handleTextChange' function handles it's onChange prop https://github.com/cerner/terra-core/pull/2800/files#diff-105fddc0e7a42eb008cb5c09faad5222R161

StephenEsser replied:
No, we'll want to stay passive with the onChange callback. We can grab a few more opinions on this before making any changes. 

My vote is `onInput(event)`.  We can passively add the searchText into the API in the future if it's requested.

BenBoersma replied:
As per offline discussion, I have switched to only returning the even on the onInput prop call https://github.com/cerner/terra-core/pull/2800/commits/d02ec4b2a3e498a93dca1c78cf449791f00389d2

supreethmr commented on packages/terra-search-field/src/terra-dev-site/doc/example/SearchFieldProgrammaticSet.jsx:
```diff
@@ -0,0 +1,81 @@
+import Button from 'terra-button/lib/Button';
+import SearchField from '../../../SearchField';
+
+class CallbackSearchField extends React.Component {
```
nit: CallbackSearchField -> SearchFieldProgrammaticSet / SearchFieldProgrammaticSearch

BenBoersma replied:
Fixed https://github.com/cerner/terra-core/pull/2800/commits/48780c9b67df1f5c0e433435270b039142582536#diff-5b54d3bd01a4e6076cc428ee0d369f6cR9 and https://github.com/cerner/terra-core/pull/2800/commits/48780c9b67df1f5c0e433435270b039142582536#diff-5b54d3bd01a4e6076cc428ee0d369f6cR9

ShettyAkarsh commented:
Found an issue like if we enter the text less than minimum required the message gets displayed saying 'Invalid Text'. this is not clearing out when we immediately add the text programmatically as shown below. 

![SearchFieldOnInput](https://user-images.githubusercontent.com/45097016/72501493-f6623780-385c-11ea-827e-b68ebe7d7d8a.gif)

Probably we might have to call handleSearch here https://github.com/cerner/terra-core/blob/f1787a51b480734950a587084eaf853e09faabfd/packages/terra-search-field/src/SearchField.jsx#L182.

supreethmr commented:
nit : 
1. spell correction in [ReadMe](https://github.com/cerner/terra-core/blob/f1787a51b480734950a587084eaf853e09faabfd/packages/terra-search-field/docs/README.md#implementation-notes)
![Screen Shot 2020-01-16 at 12 30 50 PM](https://user-images.githubusercontent.com/12869809/72525679-b619ae00-388a-11ea-9afa-3200cacabc98.png)

2. button seems to be not Aligned in example and test page : 
![Screen Shot 2020-01-16 at 11 51 31 AM](https://user-images.githubusercontent.com/12869809/72525796-faa54980-388a-11ea-8f24-65394b0d371c.png)

supreethmr commented on packages/terra-search-field/src/SearchField.jsx:
```diff
@@ -60,6 +60,11 @@ const propTypes = {
   onChange: PropTypes.func,
 
+  /**
+   * Function to trigger when user inputs a value. Sends parameters {Event} event and {String} searchText
```
Do we need to be more specific with description here like :
>  Function to trigger when user programmatically changes a value OR Function to trigger when value changes programmatically OR something more that explains the purpose this function being added in Search-field.

BenBoersma replied:
> Found an issue like if we enter the text less than minimum required the message gets displayed saying 'Invalid Text'. this is not clearing out when we immediately add the text programmatically as shown below.
> 
> ![SearchFieldOnInput](https://user-images.githubusercontent.com/45097016/72501493-f6623780-385c-11ea-827e-b68ebe7d7d8a.gif)
> 
> Probably we might have to call handleSearch here
> 
> https://github.com/cerner/terra-core/blob/f1787a51b480734950a587084eaf853e09faabfd/packages/terra-search-field/src/SearchField.jsx#L182
> 
> .

So we have two reasons we don't want handleInput to trigger the search function. 

The first reason is we don't want to double trigger the search when typing value, since onInput and onChange would fire. 

Second reason is that we don't want to give external control over when we search.

cc: @dkasper-was-taken

BenBoersma replied:
What if we wrote something like `Function to trigger when user inputs a value. Use when programmatically setting a value.` 
Though do we want to mention here that users need to manually dispatch the `input` event if they are programmatically setting?

BenBoersma replied:
Set a more verbose message here https://github.com/cerner/terra-core/commit/63069a3151a47224cf7d2c3c4ae05354b364304e#diff-105fddc0e7a42eb008cb5c09faad5222R64

BenBoersma commented:
> Found an issue like if we enter the text less than minimum required the message gets displayed saying 'Invalid Text'. this is not clearing out when we immediately add the text programmatically as shown below.
> 
> ![SearchFieldOnInput](https://user-images.githubusercontent.com/45097016/72501493-f6623780-385c-11ea-827e-b68ebe7d7d8a.gif)
> 
> Probably we might have to call handleSearch here
> 
> https://github.com/cerner/terra-core/blob/f1787a51b480734950a587084eaf853e09faabfd/packages/terra-search-field/src/SearchField.jsx#L182
> 
> .

So we have two reasons we don't want handleInput to trigger the search function. 

The first reason is we don't want to double trigger the search when typing value, since onInput and onChange would fire. 

Second reason is that we don't want to give external control over when we search.

cc: @dkasper-was-taken

BenBoersma commented:
> nit :
> 
> 1. spell correction in [ReadMe](https://github.com/cerner/terra-core/blob/f1787a51b480734950a587084eaf853e09faabfd/packages/terra-search-field/docs/README.md#implementation-notes)
>    ![Screen Shot 2020-01-16 at 12 30 50 PM](https://user-images.githubusercontent.com/12869809/72525679-b619ae00-388a-11ea-9afa-3200cacabc98.png)

Fixed https://github.com/cerner/terra-core/pull/2800/commits/48780c9b67df1f5c0e433435270b039142582536#diff-0bbd442e9dea05d7972c8517ff7ce076R27

> 2. button seems to be not Aligned in example and test page :
>    ![Screen Shot 2020-01-16 at 11 51 31 AM](https://user-images.githubusercontent.com/12869809/72525796-faa54980-388a-11ea-8f24-65394b0d371c.png)

Fixed https://github.com/cerner/terra-core/pull/2800/commits/48780c9b67df1f5c0e433435270b039142582536

yuderekyu commented on packages/terra-search-field/src/SearchField.jsx:
```diff
@@ -162,8 +167,19 @@ class SearchField extends React.Component {
   updateSearchText(searchText) {
-    if (searchText !== undefined && searchText !== this.state.searchText) {
-      this.setState({ searchText });
+    if (searchText !== undefined && searchText !== this.searchText) {
```
to cover all bases, is it worth updating this to a `typeof` check to protect from undeclared variables?
```suggestion
    if (typeof searchText !== 'undefined' && searchText !== this.searchText) {
```

yuderekyu commented on packages/terra-search-field/src/terra-dev-site/doc/example/SearchFieldProgrammaticSet.jsx:
```diff
@@ -0,0 +1,85 @@
+          bubbles: true,
+          cancelable: false,
+        });
+        customEvent.simulated = true;
```
what's the simulated for? I don't see that in the event api

BenBoersma replied:
Actually I think I left that in on accident. When I was trying to get onChange to be simulated, it was supposed to let React know that it was a "simulated" onChange event and not do the checks that prevent onChange from being programmatically fired like this.

BenBoersma replied:
Yep, totally left it in on accident, still got same behavior without it. Fixed here https://github.com/cerner/terra-core/pull/2800/commits/b86945e9e390e61f2836c59e2fe31688bc0cd737#diff-5b54d3bd01a4e6076cc428ee0d369f6cL35

yuderekyu commented on packages/terra-search-field/tests/wdio/search-field-spec.js:
```diff
@@ -134,6 +134,47 @@ Terra.describeViewports('Search Field', ['medium'], () => {
+
+    Terra.it.validatesElement('paste');
+
+    it('should not trigger search when programatticaly setting value', () => {
```
nit *programmatically

BenBoersma replied:
Fixed https://github.com/cerner/terra-core/pull/2800/commits/b86945e9e390e61f2836c59e2fe31688bc0cd737#diff-dbfac79b3a803165fdc0713a25464bcfR163

supreethmr commented:
> > Found an issue like if we enter the text less than minimum required the message gets displayed saying 'Invalid Text'. this is not clearing out when we immediately add the text programmatically as shown below.
> > ![SearchFieldOnInput](https://user-images.githubusercontent.com/45097016/72501493-f6623780-385c-11ea-827e-b68ebe7d7d8a.gif)
> > Probably we might have to call handleSearch here
> > https://github.com/cerner/terra-core/blob/f1787a51b480734950a587084eaf853e09faabfd/packages/terra-search-field/src/SearchField.jsx#L182
> > 
> > .
> 
> So we have two reasons we don't want handleInput to trigger the search function.
> 
> The first reason is we don't want to double trigger the search when typing value, since onInput and onChange would fire.
> 
> Second reason is that we don't want to give external control over when we search.
> 
> cc: @dkasper-was-taken

with reference to this example ( Added text using ref of search-field and not by using value prop of search-field hence we need to dispatch custom event for callback)  I would like to know what happens when text is added programmatically through value prop for search-field(controlled). 
Do consumer need to dispatch custom event even when value prop is changed programmatically..? 

Because i found that even with controlled search-field onChange will not be triggered when text is added programmatically. even search will not triggered until it invoked explicitly.

This could be due to component being unaware to the prop changes. so adding getDerivedStateFromProps() for search-field to update `searchText` whenever value prop is updated and calling handleSearch() would resolve this issue with controlled search-field.

```javascript
static getDerivedStateFromProps(nextProps, prevState) {
  if (nextProps.value !== prevState.prevPropsValue) {
      return {
         searchText: nextProps.value,
         prevPropsValue: nextProps.value,
      };
    }
    return null;
}

componentDidUpdate(prevProps) {
  if (prevProps.value !== this.props.value) {
      this.handleSearch();
  }
}

```

BenBoersma commented:
> with reference to this example ( Adding text using ref of search-field and not by using value prop of search-field hence we need to dispatch custom event for callback) I would like to know what happens when text is added programmatically through value prop for search-field(controlled).
> Do consumer need to dispatch custom event even when value prop is changed programmatically..?

In a controlled search-field where we change the state that is used for prop value, the re-render correctly displays the value, the clear button, and updates the internal variable. Invoking search by pressing the search button causes search-field to try to search with the new value. This has the correct functionality.

However, the current example uses the specific way in which the reported issue was programmatically changing the value. The dictation program they are using sets the value via a reference's value.

> Because i found that even with controlled search-field onChange will not be triggered when text is added programmatically. even search will not triggered until it invoked explicitly.
>
> This could be due to component being unaware to the prop changes. so adding getDerivedStateFromProps() for search-field to update `searchText` whenever value prop is updated and calling handleSearch() would resolve this issue with controlled search-field.

I do not believe we want to trigger a search when values are programmatically set.  In the use-case given to us by the issue reporter, I believe the dictation program they are using sets the value after each "word" is said, which means if we trigger search programmatically we would search on every word.

This was why we're just updating the internal variable so that an explicitly invoked search has the right value and a user could search once they are done dictating.

yuderekyu commented on packages/terra-search-field/src/terra-dev-site/doc/example/SearchFieldProgrammaticSet.jsx:
```diff
@@ -0,0 +1,84 @@
+
+const cx = classNames.bind(styles);
+
+class SearchFieldProgrammaticSet extends React.Component {
```
Not a big deal, but we should move away from class components and use functional components with hooks, whenever possible.

BenBoersma replied:
Fixed here https://github.com/cerner/terra-core/pull/2800/commits/63069a3151a47224cf7d2c3c4ae05354b364304e#diff-5b54d3bd01a4e6076cc428ee0d369f6cR1

mjhenkes commented on packages/terra-search-field/src/terra-dev-site/doc/search-field/SearchField.1.doc.jsx:
```diff
@@ -29,6 +29,8 @@ import SearchFieldFilterNumeric from '../example/SearchFieldFilterNumeric';
 import SearchFieldFocus from '../example/SearchFieldFocus';
 import SearchFieldFocusSrc from '!raw-loader!../../../../src/terra-dev-site/doc/example/SearchFieldFocus';
+import SearchFieldProgrammaticSet from '../example/SearchFieldProgrammaticSet';
+import SearchFieldProgrammaticSetSrc from '!raw-loader!../../../../src/terra-dev-site/doc/example/SearchFieldProgrammaticSet';
```
Since you're already updating this file you could switch it over to mdx

BenBoersma replied:
Switched here https://github.com/cerner/terra-core/pull/2800/commits/01caa1861a37d7efb6108958e4396e8fed923d80#diff-03a9a5c821fb01b08d491fecf3df12ccR1

PaulSolDev commented:
Functional verification complete. No issues identified.
